# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0304 Карабанов Роман, вариант 35.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования
систем компьютерной математики для реализации алгоритмов.


## Основные теоретические положения

Аддитивной цепочкой для $ n \in \mathbb{N} $ называется последовательность натуральных чисел
$$ 1 = a_0, a_1, ..., a_r = n, $$
где каждый элемент последователньости равен сумме каких-то двух предыдущих:
$$ a_i = a_j + a_k, \quad k \le j \le i, \quad i = 1, 2, ..., r $$

$l(n) = r$ - наименьшая длина аддитивной цепочки для $ n \in \mathbb{N} $.

Для метода множителей: $ \quad l(mn) \le l(m) + l(n)$

Для бинарного метода: $l(n) = \lambda(n) + \nu(n) - 1, \quad$ где $\lambda (n) = \lfloor lb(n) \rfloor, \nu(n)$ - вес Хэмминга числа $n$ (количество единиц в двоичной записи).

Для m-арного метода: $ \quad l(n) \le m - 2 + (k + 1)t, \quad$ где $m = 2^k, n = \sum_{j = 0}^t d_j m^{t-j} $

Для SX-метода: $\quad l(n) \le \lambda (n) + \nu (n) - 1$


Пара $ (j, k), 0 \le k \le j < i $ называется **шагом** $i$.Если существует более одной пары $(j, k)$, полагаем $j$ наибольшим.

## Виды шагов:
* удвоение: $ j = k = i - 1 $
* звёздный шаг: $ j = i - 1 $ (линейный шаг)
* малый шаг: $ \lambda (a_i) = \lambda (a_i - 1) $

## Свойства видов шагов:
* шаг 1 - всегда удвоение
* удвоение - звёздный шаг, но никогда не малый
* если $i$-ый шаг не малый, то $(i+1)$-ый шаг либо малый, либо звёздный, либо оба
* за удвоением всегда следует звёздный шаг
* если $(i+1)$-ый шаг не звёздный и не малый, то $i$-ый шаг должен быть малым


### Алгоритм Брауэра:

Для $n \in \mathbb{N}$ при заданном $k \in \mathbb{N}$ можно построить цепочку Брауэра с помощью рекуррентной формулы:

$$ B_k (n) =
\begin{cases}
1, 2, 3, ..., 2^k - 1, \quad n < 2^k \\
B_k (q), 2q, 4q, ..., 2^k q, n, \quad n \ge 2^k
\end{cases} \\
q = \lfloor {n \over 2^k} \rfloor
$$

Длина цепочки
$$l_B (n) = j(k + 1) + 2^k - 2,$$
при условии, что $jk \le \lambda (n) \le (j+1)k$

Длина будет минимальной для больших $n$, если положить $k = \lambda \lambda (n) - 2 \lambda \lambda \lambda (n)$

## Ход алгоритма:

* Задаётся некий параметр $k$ для $n$.
Вычисляются вспомогательные числа:
$$
d = 2^k, \hspace{0.2cm} q_1 = [ {n \over d} ], \hspace{0.2cm} r_1 = n \hspace{0.2cm} mod \hspace{0.2cm} d => n = q_1 d + r_1 \quad (0 \le r_1 < d) \\
q_2 = [ {q_1 \over d} ], \hspace{0.2cm} r_2 = q_1 \hspace{0.2cm} mod \hspace{0.2cm} d => q_1 = q_2 d + r_2 \quad (0 \le r_2 < d)
$$
* Данная процедура продолжается, пока не появится $q_s < d.$ Следовательно, $q_{s-1} = q_s d + r_s$
* Таким образом, n имеет вид
$$ n = 2^k q_1 + r_1 = 2^k (2^k q_2 + r_2) + r_1 = ...\\
... = 2^k (2^k (... (2^k q_s + r_s ) ... ) + r_2 ) + r_1 . $$

$$B_n (n): 1, 2, 3, ..., 2^k - 1, \\
2q_s, 4q_s, 8q_s, ..., 2^k q_s, 2^k q_s + r_s, \\
2q_{s-1}, 4q_{s-1}, 8q_{s-1}, ..., 2^k q_{s-1}, 2^k q_{s-1} + r_{s-1}, \\
..., \\
..., 2^k q_1, 2^k q_1 + r_1 = n.$$
  


## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных
аддитивных цепочек с использованием системы компьютерной математики
SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.
## Порядок выполнения работы
1. Вручную (т.е. не реализовывая алгоритм на Sage) построить последовательность
вычислений бинарным методом и методом множителей
для $x^n$ для 2-3 значений $n$ (значение $n > 30$ выбираются
студентом самостоятельно). Сравнить количество операций для
каждого метода, сделать выводы.
2. Реализовать алгоритм Брауэра (для нечётных вариантов) и алгоритм
Яо (для чётных вариантов) для вычисления приближённых
аддитивных цепочек для различных чисел при варьировании
параметра $k$, сопоставить длины полученных аддитивных цепочек
с минимальной аддитивной цепочкой для заданного числа. Сделать
выводы.
3. Реализовать алгоритм дробления вектора индексов для нахождения
минимальной звёздной цепочки для заданного числа. Протестировать
алгоритм минимум для 5 значений $n > 1000$. Указать,
сколько времени потребовалось на поиск цепочки и какая цепочка
получилась. Сравнить с предыдущими методами, сделать выводы.
4. Проверить гипотезу Шольца–Брауэра для всех натуральных
$1 \le n \le 12$ на алгоритме дробления вектора индексов. Результаты
оформить в виде таблицы. Сделать выводы.
5. Найти или предложить собственные модификации алгоритмов и
привести описание модификаций. Реализовать модифицированные
алгоритмы и сравнить их мощность.


## Выполнение работы
### Задание 1
####  Выберем числа n = 33, 125
#### 1.1 Бинарный метод.
Для числа n = 33 аддитивная цепочка будет выглядеть:
$$ a_0 = 1, 2, 4, 8, 16, 32, 33 = n $$
То есть для $x^n$, n = 33 число операций бинарным методом будет = 6

Для числа n = 125 аддитивная цепочка будет выглядеть:
$$ a_0 = 1, 2, 3, 6, 7, 14, 15, 30, 31, 62, 124, 125 = n$$
То есть для $x^n$, n = 64 число операций бинарным методом будет = 11

#### 1.1 Метод множителей.
Для числа n = 33 

$x^{33} = (x^{11})^3, (y^3 = y^2 * y, y^2 = y * y)$

$x^{11} = x^{10} * x$

$x^{10} = (x^5)^2$

$x^5  = x^4 * x$

$x^4 = x^2 * x^2$

$x^2 = x * x$

Отсюда требуемые операции : $ x, x^2, x^4, x^5, x^{10}, x^{11}, x^{22}, x^{33} $ = 7

Для числа  = 125

$x^{125} = (x^{25})^5, (y^5 = y^4 * y, y^4 = y^2 * y^2, y^2 = y * y)$

$x^{25} = (x^5)^5, (y^5 = y^4 * y, y^4 = y^2 * y^2, y^2 = y * y)$

$x^5 = x^4 * x $

$x^4 = x^2 * x^2 $

$x^2 = x * x$

Отсюда требуемые операции : $ x, x^2, x^4, x^5, x^{10}, x^{20}, x^{25}, x^{50}, x^{100}, x^{125} $ = 9

### Вывод на основе результов используемых значений
Таким образом видно, что для числа 33 метод множителей даёт результат по операция больший, чем бинарный метод, т.к он работает "медленне" по отношению к бинарному методу при степенях $ n =  2^k + 1 $ в то время, как для числа 125 метод множителей даёт меньшее количество операций, по сравнению с бинарным методом. Можно сделать выводы о том, что количество операций, необходимых для метода множителей равно сумме длин аддитивных цепочек сомножителей, поэтому ни один из методов не может рассматриваться как более эффективный.

### Задание 2
### Реализация алгоритма Брауэра
Реализованный алгоритм Брауэра представлен, с вариацией значений k представлен ниже:

In [2]:
def alg(n, k):
      d = 2 ** k
      ans = []
      if (n < d):
        i = 0
        while i < n:
            ans.append(i + 1)
            i += 1
      else:
          q = n
          ans_r = []
          while q >= d:
              r = q % d
              ans_r.append(r)
              q = q // d
          ans_r.append(q)
          ans_r.reverse()
          z = 1
          while z < d:
              ans.append(z)
              z += 1          
          q = ans_r[0]
          i = 0
          while i < (len(ans_r) - 1):
              j = 0
              while j < k:
                  q *= 2
                  ans.append(q)
                  j += 1
              if ans_r[i+1] > 0:
                  q += ans_r[i+1]
                  ans.append(q)
              i += 1
      return ans, (len(ans))

n = int(input("n: "))
k = int(input("k: "))
ad = []
lng = 0
ad, lng = alg(n, k)
print("Аддитивная цепочка: ", ad)
print("Длина аддитивной цепочки: ", lng)


n: 1023
k: 1
Аддитивная цепочка:  [1, 2, 3, 6, 7, 14, 15, 30, 31, 62, 63, 126, 127, 254, 255, 510, 511, 1022, 1023]
Длина аддитивной цепочки:  19


Для $ n = 90 $ результаты:

| n  | k | минимальная длина | длина | аддитивная цепочка                                 |
|----|---|-------------------|-------|----------------------------------------------------|
| 90 | 2 | 9                 | 12    | 1,2,3,4,5,10,20,22,44,88,90                        |
| 90 | 3 | 9                 | 15    | 1,2,3,4,5,6,7,2,4,8,11,22,44,88,90                 |
| 90 | 4 | 9                 | 20    | 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,10,20,40,80,90 |

Для $ n = 512$ результаты:

| n  | k | минимальная длина | длина | аддитивная цепочка                                 |
|----|---|-------------------|-------|----------------------------------------------------|
| 512 | 2 | 11                 | 11    | 1, 2, 3, 4, 8, 16, 32, 64, 128, 256, 512                       |
| 512 | 3 | 11                 | 16    | 1, 2, 3, 4, 5, 6, 7, 2, 4, 8, 16, 32, 64, 128, 256, 512                 |
| 512 | 4 | 11                 | 20    | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 4, 8, 16, 32, 64, 128, 256, 512 |

Для $ n = 127 $ результаты:

| n  | k | минимальная длина | длина | аддитивная цепочка                                 |
|----|---|-------------------|-------|----------------------------------------------------|
| 127 | 2 | 12                 | 12    | 1, 2, 3, 2, 4, 7, 14, 28, 31, 62, 124, 127                        |
| 127 | 3 | 12                 | 15    | 1, 2, 3, 4, 5, 6, 7, 2, 4, 8, 15, 30, 60, 120, 127              |
| 127 | 10 | 12                 | 127    | 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, ... , 127 |

Таким образом, можно сделать вывод, что алгоритм Брауэра является приближенным и при определённых значениях $ k $ выдаёт цепочки близкие или совпадающие с минимальными, но при больших k длина аддитивной становится во много раз больше минимальной (последний пример)


### Задание 3
### Алгоритм дробления вектора индексов


Опишем функции $ \nu(n) $ для уточнения отрезков рассматриваемых длин цепочек:

In [ ]:
def nu(n):
    res = 0
    while n != 0:
        if n % 2 == 1:
            res += 1
        n //= 2

    return res

Далее опишем сам алгоритм дробления индексов и по нему построим минимальную звёздную цепочку

In [ ]:
import time

def get_a(vec): 
    additive_chain = []
    additive_chain.append(1)
    for i in range(len(vec)):
        additive_chain.append(additive_chain[i] + additive_chain[vec[i] - 1])
    
    return additive_chain[-1]

def reduce(vec):
    _len = len(vec)
    for i in range(_len):
        if vec[_len - i - 1] > 1:
            vec[_len - i - 1] -= 1
            for j in range(_len - i, _len):
                vec[j] = j + 1
            break

def get_star_chain(star_chain_vec):
    additive_chain = [1]
    for i in range(len(star_chain_vec)):
        additive_chain.append(additive_chain[i] + additive_chain[star_chain_vec[i] - 1])
    
    return additive_chain

def not_reduce(vec):
    for i in vec:
        if i > 1:
            return False
    
    return True

def splitting_alg(n):
    lower_border = math.ceil(math.log(n, 2))
    upper_border = int(math.log(n, 2) + nu(n))
    for m in range(lower_border, upper_border):
        q = int(m/2)
        fix_path = [i for i in range(1, q + 1)]
        while True:
            change_path = [q + i for i in range(1, m - q + 1)]
            a = get_a(fix_path)
            a_min = a + m - q
            a_max = a * 2 ** (m - q)

            if n < a_min or n > a_max:
                if not_reduce(fix_path) and len(fix_path) > 1:
                    break
                reduce(fix_path)
                continue
                
            while True: 
                if get_a(fix_path + change_path) == n:
                    return fix_path + change_path
                if not_reduce(change_path):
                    break
                reduce(change_path)
            if not_reduce(fix_path):
                break
            reduce(fix_path)   

n = int(input())
start = time.time()
star_chain_vec = splitting_alg(n)
end = time.time()
print("Звёздная цепочка : ", get_star_chain(star_chain_vec))
print("Длина зцепочки : ", len(get_star_chain(star_chain_vec)))
print("Время выполнения : ", end - start)

Результаты работы алгоритма дробления индексов для $n = 1028, 1548, 1601, 1672, 1802$ представлены в таблице:

| n | Длина цепочки | Цепочка | Время поиска, с |
| --|:-------------:|:--------:|:-----------------:|
| 1048 | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028]| 0.002196 |
| 1548 | 13 |[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 516, 1032, 1548]| 2.8064 |
| 1601 | 14 |[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1600, 1601]| 566.1104 |
| 1672 | 14 |[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1536, 1664, 1672]|  439.2508 |
| 1802 | 14 |[1, 2, 4, 8, 16, 32, 64, 128, 256, 258, 516, 772, 1544, 1802] | 292.3192 |


### Задание 4
### Проверить гипотезу Шольца–Брауэра для всех натуральных  $1≤𝑛≤12$  на алгоритме дробления вектора индексов.



Гипотеза Штольца - Брауэра заключается в: $ l(2^n - 1) \le l(n) + n - 1 $

Проверим её для $n \in [1, 12]$

In [ ]:
for i in range(1, 13):
    right = splitting_alg(i)
    left = splitting_alg(2**i - 1)
    right_chain = get_star_chain(right)
    left_chain = get_star_chain(left)
    print(i, " ", len(left_chain), " <= ", len(right_chain) + i - 1," : ", len(left_chain) <= i - 1 + len(right_chain))

Вывод:
1   1  <=  1  :  True
2   3  <=  3  :  True
3   5  <=  5  :  True
4   6  <=  6  :  True
5   8  <=  8  :  True
6   9  <=  9  :  True
7   11  <=  11  :  True
8   11  <=  11  :  True
9   13  <=  13  :  True
10   14  <=  14  :  True
11   16  <=  16  :  True
12   16  <=  16  :  True

| n           | $ l(2^n - 1) $| $l(n) + n - 1 $ |
|-------------|---------------|-----------------|
| 1 | 1             | 1               | 
| 2 | 3             | 3               |
| 3 | 5             | 4               |
| 4 | 6             | 5               |
| 5 | 8             | 8               |
| 6 | 9             | 9               |
| 7 | 11            | 11              | 
| 8 | 11            | 11              | 
| 9 | 13            | 13              | 
| 10| 14            | 14              | 
| 11| 16            | 16              | 
| 12| 16            | 16              | 

На основе данных можно сделать вывод, что гипотеза Шольца-Брауэра справедлива для чисел $n \in [1, 12]$

## Вывод

В ходе выполнения практической работы были реализованы алгоритм Брауэра, алгоритм дробления вектора индексов и проверена гипотеза Шольца-Брауэра для $n \in [1, 12]$. Для алгоритма Брауэра можно сделать вывод, что он строит цепочку близкую к минимальной для малых $ k $, относительно больших чисел. Алгоритм дробления индексов строит минимальную звёздную цепочку для заданного числа, но время его работы достаточно велико. Гипотеза Шольца-Брауэра была проверена на алгоритме дробления вектора индексов.